### PR TITLE
mem: QoS q_policy assertions fix

### PR DESCRIPTION
**Pull request by hristobelchev**
Fix QoS Memory Queue Policies

* Fix assertions in LRG policy to correctly assert requestor and list validity
* Fix `selectPacket()` in LIFO Queue Policy to correctly return the end of the `deque` backing store for its packet queue

### CODE REVIEW

**Review by powerjg — DISMISSED** [review]

File: src/mem/qos/q_policy.hh [hunk 1]
```diff
@@ -127,13 +127,15 @@ class LifoQueuePolicy : public QueuePolicy
...
      *
-     * @param queue The queue in which to select a packet
+     * @param queue The non-empty queue from which to select a packet
      * @return Iterator to the selected packet
      */
     PacketQueue::iterator
     selectPacket(PacketQueue* queue) override
     {
-        return queue->end();
+        panic_if(queue->empty(),
```
Comment (odanrc) [inline src/mem/qos/q_policy.hh:136]: include base/logging.hh. Better option: move this implementation to .cc file

File: src/mem/qos/q_policy.hh [hunk 2]
```diff
@@ -148,12 +150,14 @@ class FifoQueuePolicy : public QueuePolicy
...
      * Implements FCFS packet select policy
      *
-     * @param queue The queue in which to select a packet
+     * @param queue The non-empty queue from which to select a packet
      * @return Iterator to the selected packet
      */
     PacketQueue::iterator
     selectPacket(PacketQueue* queue) override
     {
+        panic_if(queue->empty(),
```
Comment (odanrc) [inline src/mem/qos/q_policy.hh:159]: Ditto

File: src/mem/qos/q_policy.cc [hunk 3]
```diff
@@ -72,6 +72,8 @@ QueuePolicy::create(const QoSMemCtrlParams &p)
 QueuePolicy::PacketQueue::iterator
 LrgQueuePolicy::selectPacket(PacketQueue* q)
 {
+    panic_if(q->empty(),
```
Comment (odanrc) [inline src/mem/qos/q_policy.cc:75]: The comment on line 142 probably needs an update

**Review by odanrc — APPROVED** [review]

File: src/mem/qos/q_policy.cc (hunk 3 above)
Comment (hristobelchev) [inline src/mem/qos/q_policy.cc:91]: Thanks, I hadn't noticed it - I've updated the comment to reflect the changes

**Review by powerjg — APPROVED** [review]

**Pull request merged**